File: lib/wallace/schemas/transaction.ex
defmodule Wallace.Schemas.Transaction do
  use Ecto.Schema
  use Wallace.Schemas.Base

  import Ecto.Changeset

  alias Wallace.Schemas.Account
  alias Wallace.Schemas.Category
  alias Wallace.Schemas.Payee
  alias Wallace.Schemas.TransactionType
  alias Wallace.Schemas.Types.Repetition

  schema "transactions" do
    field :amount, Money.Ecto.Composite.Type
    field :date, :utc_datetime
    field :notes, :string
    field :repetition, Repetition

    belongs_to :account, Account
    belongs_to :category, Category
    belongs_to :payee, Payee
    belongs_to :type, TransactionType

    timestamps()
  end

  @doc false
  def changeset(transaction, attrs) do
    transaction
    |> cast(attrs, [
      :amount,
      :date,
      :notes,
      :repetition,
      :account_id,
      :category_id,
      :payee_id,
      :type_id
    ])
    |> validate_required([
      :amount,
      :date,
      :notes,
      :repetition,
      :account_id,
      :category_id,
      :type_id
    ])
    |> validate_min_money
  end

  defp validate_min_money(changeset) do
    validate_change(changeset, :amount, fn
      _, %Money{amount: amount} when amount > 0 -> []
      _, _ -> [amount: "must be greater than 0"]
    end)
  end
end

File: wallace/Domain/Entities/Transaction.cs
using System;
using NodaMoney;
using Wallace.Domain.Enums;

namespace Wallace.Domain.Entities
{
    /// <summary>
    /// Represents a transaction of a certain account.
    /// </summary>
    public class Transaction : IOwnedEntity
    {
        public Guid Id { get; set; }
        public TransactionType Type { get; set; }
        public Repetition Repetition { get; set; }
        public Money Amount { get; set; }
        public DateTime Date { get; set; }
        public string Notes { get; set; }

        public Guid OwnerId { get; set; }
        public User Owner { get; set; }
        
        public Guid AccountId { get; set; }
        public Account Account { get; set; }
        
        public Guid? CategoryId { get; set; }
        public Category Category { get; set; }
        
        public Guid? PayeeId { get; set; }
        public Payee Payee { get; set; }
    }
}